integration/pkg/api/osrm/table/request.go

```diff
@@ -0,0 +1,159 @@
+package table
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/Telenav/osrm-backend/integration/pkg/api"
+	"github.com/Telenav/osrm-backend/integration/pkg/api/osrm/coordinate"
+	"github.com/Telenav/osrm-backend/integration/pkg/api/osrm/genericoptions"
+	routeoption "github.com/Telenav/osrm-backend/integration/pkg/api/osrm/route/options"
+	"github.com/Telenav/osrm-backend/integration/pkg/api/osrm/table/options"
+	"github.com/golang/glog"
+)
+
+// Request for OSRM table service
+// http://project-osrm.org/docs/v5.5.1/api/#table-service
+type Request struct {
+	// Path
+	Service     string
+	Version     string
+	Profile     string
+	Coordinates coordinate.Coordinates
+
+	// Options
+	Sources      genericoptions.Elements
+	Destinations genericoptions.Elements
+	Annotations  string
+}
+
+// NewRequest create an empty table Request.
+func NewRequest() *Request {
+
+	return &Request{
+		// Path
+		Service:     "table",
+		Version:     "v1",
+		Profile:     "driving",
+		Coordinates: coordinate.Coordinates{},
+
+		// Options
+		Sources:      genericoptions.Elements{},
+		Destinations: genericoptions.Elements{},
+		Annotations:  "",
+	}
+
+}
+
+// ParseRequestURL parse Request URL to Request.
+func ParseRequestURL(u *url.URL) (*Request, error) {
+	if u == nil {
+		return nil, fmt.Errorf("empty URL input")
+	}
+
+	req := NewRequest()
+
+	if err := req.parsePath(u.Path); err != nil {
+		return nil, err
+	}
+
+	req.parseQuery(api.ParseQueryDiscardError(u.RawQuery))
+
+	return req, nil
+}
+
+func (r *Request) parsePath(path string) error {
+	p := path
+	p = strings.TrimPrefix(p, api.Slash)
+	p = strings.TrimSuffix(p, api.Slash)
+
+	s := strings.Split(p, api.Slash)
+	if len(s) < 4 {
+		return fmt.Errorf("invalid path values %v parsed from %s", s, path)
+	}
+	r.Service = s[0]
+	r.Version = s[1]
+	r.Profile = s[2]
+
+	var err error
+	if r.Coordinates, err = coordinate.ParseCoordinates(s[3]); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *Request) parseQuery(values url.Values) {
+
+	if v := values.Get(options.KeySources); len(v) > 0 {
+		if sources, err := genericoptions.ParseElemenets(v); err == nil {
+			r.Sources = sources
+		}
+	}
+
+	if v := values.Get(options.KeyDestinations); len(v) > 0 {
+		if destinations, err := genericoptions.ParseElemenets(v); err == nil {
+			r.Destinations = destinations
+		}
+	}
+
+	if v := values.Get(options.KeyAnnotations); len(v) > 0 {
+		if annotations, err := routeoption.ParseAnnotations(v); err == nil {
```

Comment:
The `annotations` of `route` service are different from that of `table` service. `table` service should have its own parser.